### PR TITLE
[feat]: [SRM-14726]: write to CompositeSLORecordBucket in addition to write to CompositeSLORecord

### PR DESCRIPTION
**Pull request by n00bitax**
## Describe your changes

- In addition to writing to CompositeSLORecord , records are also written to CompositeSLORecordBucket in chunk size of 5 currently.
- CompositeSLORecordBucket reads from the SLIRecordBucket only and all windows and processing takes on multiples of SLI_RECORD_BUCKET_SIZE only.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55395)

### CODE REVIEW

**Comment by n00bitax** [comment]
trigger uts

**Review by deepak-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55395)* status: 13 file(s) to review. 0 of 1 approvals obtained. 

___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java` line 74 at r1](https://reviewable.io/reviews/harness/harness-core/55395#-NjQeOdKErXY2OtqKExu:-NjQeOdLDqfism2nIxjN:b1ufw7y) ([raw file](https://github.com/harness/harness-core/blob/d7ca8ba58cab5e3ca0a75dbd2193579092a9257a/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java#L74)):*
> ```Java
>   @Id private String uuid;
>   @FdIndex private String verificationTaskId;
>   private String sloId;
> ```

We can remove one of the verification task id or sloId.
I will say sloId. Since verification task id helps in code cleanup(unless we are ready to change that logic).
___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java` line 76 at r1](https://reviewable.io/reviews/harness/harness-core/55395#-NjQe9Xp89zTM76Rob0-:-NjQe9Xp89zTM76Rob00:b-sg3iog) ([raw file](https://github.com/harness/harness-core/blob/d7ca8ba58cab5e3ca0a75dbd2193579092a9257a/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java#L76)):*
> ```Java
>   private String sloId;
>   private Instant startTimestamp; // minute
>   @Setter(AccessLevel.PRIVATE) private long epochMinute;
> ```

we can drop epochMinute. I don't think we use it. If it was added in SLIBucketRecord we should remove from there as well @Varsha-lalwani 
___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/api/SLIRecordBucketService.java` line 53 at r1](https://reviewable.io/reviews/harness/harness-core/55395#-NjQed2t85ePb0Z35Op1:-NjQed2t85ePb0Z35Op2:b-vdn57p) ([raw file](https://github.com/harness/harness-core/blob/d7ca8ba58cab5e3ca0a75dbd2193579092a9257a/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/api/SLIRecordBucketService.java#L53)):*
> ```Java
>       SLIRecordBucket sliRecordBucket, ServiceLevelIndicator serviceLevelIndicator);
> 
>   org.apache.commons.math3.util
> ```

We should not have complete class path. 
___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java` line 78 at r1](https://reviewable.io/reviews/harness/harness-core/55395#-NjQf9aLDRulW9ukL9JC:-NjQf9aLDRulW9ukL9JD:b-fu5lnq) ([raw file](https://github.com/harness/harness-core/blob/d7ca8ba58cab5e3ca0a75dbd2193579092a9257a/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java#L78)):*
> ```Java
>   }
> 
>   private Instant getStartTime(CompositeServiceLevelObjective compositeServiceLevelObjective) {
> ```

Good one. But if we make it a AbstractSLO function only it will be great. We have this code scatttered at multiple places in the codebase.

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/impl/SLIRecordBucketServiceImpl.java [hunk 1]
```diff
@@ -326,4 +333,66 @@ public Pair<Long, Long> getPreviousBucketRunningCount(
...
+          simpleServiceLevelObjective.getServiceLevelIndicators().get(0));
+      String sliId = serviceLevelIndicator.getUuid();
+      int sliVersion = serviceLevelIndicator.getVersion();
+      if (serviceLevelIndicator.getSLIEvaluationType().equals(SLIEvaluationType.WINDOW)
+          && serviceLevelIndicator.getConsiderConsecutiveMinutes() != null
+          && serviceLevelIndicator.getConsiderConsecutiveMinutes() > 1) {
+        SLIRecordBucket lastSLIRecordBucket = getLatestSLIRecordSLIVersion(sliId, sliVersion);
+        if (lastSLIRecordBucket != null) {
+          // TODO check again, we are adjusting endtime here
+          Instant timeOfLastRecordWhichIsFixed = lastSLIRecordBucket.getBucketStartTime().minus(
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/impl/SLIRecordBucketServiceImpl.java:375]: will also need to round this down to nearest 5 minutes

File: srm-service/modules/cv-nextgen-service/service/src/test/java/io/harness/cvng/servicelevelobjective/services/impl/CompositeSLORecordBucketServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,1077 @@
...
+    List<CompositeSLORecordBucket> sloRecords1 =
+        sloRecordService.getSLORecordBuckets(verificationTaskId, startTime, endTime);
+    assertThat(sloRecords1.size()).isEqualTo(1);
+    assertThat(sloRecords1.get(0).getRunningBadCount()).isEqualTo(0.0);
+    assertThat(sloRecords1.get(0).getRunningGoodCount()).isEqualTo(0.0);
+    assertThat(sloRecords1.get(0).getSloVersion()).isEqualTo(0);
+  }
+
+  @Test
+  @Owner(developers = VARSHA_LALWANI)
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/test/java/io/harness/cvng/servicelevelobjective/services/impl/CompositeSLORecordBucketServiceImplTest.java:604]: can you change the owner to yourself

**Review by Varsha-lalwani — CHANGES_REQUESTED** [review]

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger messagecheck

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/test/java/io/harness/cvng/servicelevelobjective/services/impl/CompositeSLORecordBucketServiceImplTest.java (hunk 2 above)
Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/test/java/io/harness/cvng/servicelevelobjective/services/impl/CompositeSLORecordBucketServiceImplTest.java]: done

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/impl/SLIRecordBucketServiceImpl.java (hunk 1 above)
Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/impl/SLIRecordBucketServiceImpl.java]: done

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      CompositeSLORecordBucket compositeSLORecordBucket = unsafeBuild();
+      compositeSLORecordBucket.setEpochMinute(TimeUnit.MILLISECONDS.toMinutes(startTimestamp.toEpochMilli()));
+      return compositeSLORecordBucket;
+    }
+  }
+  @Version long version;
+  @Id private String uuid;
+  @FdIndex private String verificationTaskId;
+  private String sloId;
+  private Instant startTimestamp; // minute
```
[the remark this answers is not in the document]
Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java]: done

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

**Review by Varsha-lalwani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55395)* status: 18 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java` line 76 at r1](https://reviewable.io/reviews/harness/harness-core/55395#-NjQe9Xp89zTM76Rob0-:-NjkHT27Dm0xlIhhuNo_:bdzlbjb) ([raw file](https://github.com/harness/harness-core/blob/d7ca8ba58cab5e3ca0a75dbd2193579092a9257a/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/entities/CompositeSLORecordBucket.java#L76)):*
<details><summary><i>Previously, deepak-harness (Deepak Chhikara) wrote…</i></summary><blockquote>

we can drop epochMinute. I don't think we use it. If it was added in SLIBucketRecord we should remove from there as well @Varsha-lalwani
</blockquote></details>

yes we should.

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

**Review by Varsha-lalwani — APPROVED** [review]

**Comment by Varsha-lalwani** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger ut5,ut9

**Comment by n00bitax** [comment]
trigger ut5,ut6,ut7,ut8,ut9

**Comment by n00bitax** [comment]
trigger ut6,ut7,ut9

**Comment by n00bitax** [comment]
trigger ut7,ut8,ut9

**Pull request merged**